RoslynPluginGenerator/ArchiveUpdater.cs

```diff
@@ -0,0 +1,154 @@
+        private readonly IDictionary<string, string> fileMap;
+        private readonly ILogger logger;
+
+        private string inputArchiveFilePath;
+        private string outputArchiveFilePath;
+
+        #region Public methods
+
+        public ArchiveUpdater(string workingDirectory, ILogger logger)
+        {
+            if (string.IsNullOrWhiteSpace(workingDirectory))
+            {
+                throw new ArgumentNullException("workingDirectory");
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException("logger");
+            }
+            this.logger = logger;
+            this.workingDirectory = workingDirectory;
+
+            this.fileMap = new Dictionary<string, string>();
+        }
+
+        public ArchiveUpdater SetInputArchive(string filePath)
+        {
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                throw new ArgumentNullException("filePath");
+            }
+            this.inputArchiveFilePath = filePath;
+            return this;
+        }
+
+        public ArchiveUpdater SetOutputArchive(string filePath)
+        {
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                throw new ArgumentNullException("filePath");
+            }
+            this.outputArchiveFilePath = filePath;
+            return this;
+        }
+
+        public ArchiveUpdater AddFile(string sourceFilePath, string relativeTargetFilePath)
+        {
+            this.fileMap[relativeTargetFilePath] = sourceFilePath;
+            return this;
+        }
+
+        public void UpdateArchive()
+        {
+            this.logger.LogDebug(UIResources.ZIP_UpdatingArchive, this.inputArchiveFilePath);
+
+            string unpackedDir = Utilities.CreateSubDirectory(this.workingDirectory, "unpacked");
+            this.logger.LogDebug(UIResources.ZIP_WorkingDirectory, workingDirectory);
+
+            ZipFile.ExtractToDirectory(this.inputArchiveFilePath, unpackedDir);
+
+            // Add in the new files
+            foreach (KeyValuePair<string, string> kvp in this.fileMap)
+            {
+                this.logger.LogDebug(UIResources.ZIP_InsertingFile, kvp.Key, kvp.Value);
+
+                string targetFilePath = Path.Combine(unpackedDir, kvp.Key);
+                Directory.CreateDirectory(Path.GetDirectoryName(targetFilePath));
+                File.Copy(kvp.Value, targetFilePath);
+            }
+
+            // Re-zip
+            if (File.Exists(this.outputArchiveFilePath))
+            {
+                this.logger.LogDebug(UIResources.ZIP_DeletingExistingArchive);
+                File.Delete(this.outputArchiveFilePath);
+            }
+
+            ZipUsingShell(unpackedDir, this.outputArchiveFilePath);
+
+            this.logger.LogDebug(UIResources.ZIP_NewArchiveCreated, this.outputArchiveFilePath);
+        }
+        
+        private static void ZipUsingShell(string sourceDir, string targetZipFilePath)
```

Comment:
This method does effectively the same as the PowerShell script in the MSBuild scanner packaging project i.e. uses a COM shell class to re-zip the folder (using the .Net ZipFile class creates an invalid jar).
I'll add a pointer to the Scanner PowerShell script and add a comment to this method.